### PR TITLE
feat: typescript utilities

### PR DESCRIPTION
**Pull request by tien**
- Improve structures type definitions
```ts
type Person = Vertex<"person", { name: string; age: number }>;
type Pet = Vertex<"pet", { name: string; age: number }>;
type Owns = Edge<Person, "owns", Pet, { since: Date }>;
```

### CODE REVIEW

**Comment by FlorianHockmann** [comment]
If I'm not mistaken, then this PR does two things which are mostly separate from each other, right? It extends the type definitions for existing types like `Vertex`, `Edge`, and `VertexProperty` and it also adds a new utility method `parseVertex` which can be used to create a `Vertex` object where a `VertexProperty` is directly a property of the object, e.g., `vertex.name`.

I think it would be good to separate these two things from each other. Could you please create issues for them in our JIRA? Our usual development workflow is to first have an issue there which describes the bug / proposed improvement and afterwards create the PR. This also makes it possible to first discuss the issue before actually implementing it.

In this case, I'm not 100% about the added utility method. It sounds to me like it's a first step in the direction of the [_Sugar Plugin_](https://tinkerpop.apache.org/docs/current/reference/#sugar-plugin) for Groovy and I'm not sure whether we should implement such syntactic sugar for the GLVs like gremlin-javascript.
But maybe I'm just missing context why this is useful for Javascript / Typescript. You could probably better describe your motivation for this in the JIRA issue.

Sorry for the slight push back here in general. Your Typescript improvements are more than welcome, but I think they are difficult for us to review since we simply lack some experience in this area. That is why I think that following this process helps as it makes it easier for other contributors to better understand the improvements before reviewing them.

**Comment by Cole-Greer** [comment]
Hi @tien, thanks for putting this together.

I agree with Florian that for this type of change, it is best to start with a JIRA or dev list post before starting implementation.

I do think that consistency between the GLV's is important, but I can also see how this could be a valuable convenience for users. I do think if we are to add this for vertices in JS, we need to do the equivalent for edges.

In terms of naming, an alternative worth considering might be `flattenVertex` but this isn't something I feel strongly about.

The overall implementation looks good to me, although I wonder if the need to provide explicit cardinality schemas for anything other than singles will limit the convenience upside here. It would be cool if there was an option to try and detect the cardinality of each property, and only force a conversion if the user explicitly specifies an override. Granted this would currently be limited to just single and list, as set is currently blocked by [TINKERPOP-3011](https://issues.apache.org/jira/browse/TINKERPOP-3011)

In general, if this is something that JS users want in the GLV, then I'm willing to put aside some of the cross-GLV consistency concerns here. I think it's worth a quick dev list thread just to ensure the design is aligned with the community.

**Comment by tien** [comment]
Hey guys, thanks for the feedbacks, yes that make sense. So I've reduced the scope of the PR change to just improving the TypeScript definition for graph structures: vertex, edge, etc

**Comment by tien** [comment]
I mainly included the utility function as a demonstration of what a strong TypeScript definition can enable.

As for the benefit of having an utility function like that, with TypeScript I have to do a lot of boilerplates just extract basic values from vertices, i.e:

```ts
const vertices = await g.V().hasLabel("person").toList<PersonVertex>();

const people = vertices.map(vertex => {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const age = vertex.properties.age.at(0)!.value;

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const name = vertex.properties.name.at(0)!.value;
  
  const addresses = vertex.properties.address.map(x => x.value);

  return {
    age,
    name,
    addresses
  };
});
```

which with the utility function can be simplified to

```ts
const vertices = await g.V().hasLabel("person").toList<PersonVertex>();
const people = vertices.map(vertex => parseVertex(vertex, { address: "list" }));
```

**Comment by tien** [comment]
I'm happy to pursue this with a Jira ticket or via the dev list as mentioned if you guys think a utility like this is valuable to be included in the core gremlin lib.

**Comment by vkagamlyk** [comment]
Hi @tien,
I think this is a good direction for improvement for TinkerPop 5/6 when/if we introduce schema, but now it is important to have similar capabilities in all GLV's.

Perhaps it makes sense to make an example from this code? Or some utility class?

this is really beautiful code, but my opinion is VOTE+0

**Comment by tien** [comment]
Hey @vkagamlyk, as per my comment [here](https://github.com/apache/tinkerpop/pull/2599#issuecomment-2118831749) I've removed all none GLV's feature parity changes, so this PR will just be improving the TypeScript definition. No feature changes are involved.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/io/type-serializers.ts
Comment (FlorianHockmann) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/io/type-serializers.ts:352]: Can `outV` really be `undefined`? And shouldn't users get an exception here if it is?

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts
Comment (FlorianHockmann) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts:81]: Why `TOutVertex extends Vertex`? Why not simply `Vertex`?

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts
Comment (FlorianHockmann) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts:85]: I think this is too restrictive. Providers can use any type they like for the ID. JanusGraph for example has its own type.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts
Comment (FlorianHockmann) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts:142]: Sorry, but I don't know this notation and it's a bit hard to find any information on it as I don't even know what to search for here 😅 
Would you be so kind to provide me with some documentation for this kind of type definition or maybe the name of the concept for this so I can search for it myself? :-)

**Review by FlorianHockmann — COMMENTED** [review]
I just tried to review this, but I have to say that my TypeScript knowledge is quite limited. So most of my comments are really only questions of understanding.

In some cases, your type definitions are also more strict than what we have in other GLVs, and especially in Java which should be the reference.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts
Comment (tien) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts:48]: In TypeScript, we have string literal, which can be a generic, example:

```ts
type Foo<T> = T;

type Bar = Foo<"string-literal">;
```

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts
Comment (tien) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts:81]: The extends part will restrict the input type to `Vertex`, while the `= Vertex` part will default that generic to a generic Vertex if no type is provided.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts
Comment (tien) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts:142]: I believe you are referring to this part?

```ts
[VertexProperty<TLabel, TValue, TProperties, TId>, ...Array<VertexProperty<TLabel, TValue, TProperties, TId>>]
```

It basically will ensure the array will have at least 1 element.

You can find more information regarding it [here](https://www.typescriptlang.org/docs/handbook/variable-declarations.html#spread) :D

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts
Comment (tien) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts:85]: Thanks, I've removed the type restriction.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/io/type-serializers.ts
Comment (tien) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/io/type-serializers.ts]: Oops, nice catch, I've removed the unnecessary optional chaining.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts
Comment (joshsh) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts:61]: This looks like good work, but I don't know TypeScript. Quick question to get oriented: is optionality in property types supported, and if so, how? I.e. can you declare a certain vertex type to have a mandatory string-valued "name" property, and another vertex type to have an optional string-valued "name" property?

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts
Comment (joshsh) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts:48]: I'm not sure I follow this, either, but I want to understand. What is the difference between string-literal and string? I think of both vertex labels and edge labels as simply being strings, or being of a special vertex-label or edge-label type which contains string.

Comment (tien) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts:48]: TypeScript string literal are kinda like enum in Java or C#, so for example:

```ts
type LiteralPerson = Vertex<"person", { name: string; age: number }>;
type LoosePerson = Vertex<string, { name: string; age: number }>;

let person1: LiteralPerson // person1.label will have a const type of `"person"`
let person2: LoosePerson // person2.label will instead only be typed as `string`
```

source: https://www.typescriptlang.org/docs/handbook/literal-types.html#string-literal-types

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts
Comment (tien) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts:61]: Yes, you can do something like this:

```ts
type Human = Vertex<"person", { name: string; }>;
type NewBorn = Vertex<"person", { name?: string; }>;
```

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts
Comment (FlorianHockmann) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts:48]: I see, so users who don't define their own types (like `Person`) will simply have a `Vertex` where `label` is a `string`, but users can restrict the `label` to their own allowed set of label if they want to.
Thanks for the explanation! I'm marking this as resolved now. @joshsh feel free to unresolve if you more questions about this.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts
Comment (FlorianHockmann) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/graph.ts:142]: Interesting, didn't know the spread operator. Thanks for the explanation :)

**Review by FlorianHockmann — APPROVED** [review]
Thanks a lot for your explanations and your patience here, @tien!

VOTE +1

**Comment by xiazcy** [comment]
Thank you for the contribution! As lazy consensus has been reached, I'll be merging this PR today if there are no further comments. 

One small note, it does look like this PR is missing a CHANGELOG entry, which I'm happy to add it after merging as CTR.

**Pull request merged**
